lib/routes/recovery-keys.js

```diff
@@ -0,0 +1,230 @@
+            recoveryKeyId: isA.string().max(32),
+            recoveryData: isA.string().max(1024)
+          }
+        }
+      },
+      handler(request, reply) {
+        log.begin('createRecoveryKey', request)
+
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const {recoveryKeyId, recoveryData} = request.payload
+
+        customs.check(request, 'createRecoveryKey')
+          .then(createRecoveryKey)
+          .then(emitMetrics)
+          .then(() => reply({}), reply)
+
+        function createRecoveryKey() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.createRecoveryKey(uid, recoveryKeyId, recoveryData)
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'account.recoveryKey.created',
+            uid: uid
+          })
+
+          return request.emitMetricsEvent('recoveryKey.created', {uid: uid})
+            .then(() => ({}))
+        }
+      }
+    },
+    {
+      method: 'GET',
+      path: '/recoveryKeys/{recoveryKeyId}',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        }
+      },
+      handler(request, reply) {
+        log.begin('getRecoveryKey', request)
+
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const recoveryKeyId = request.params.recoveryKeyId
+        let recoveryData
+
+        customs.check(request, 'getRecoveryKey')
+          .then(getRecoveryKey)
+          .then(() => reply({recoveryData}), reply)
+
+        function getRecoveryKey() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.getRecoveryKey(uid, recoveryKeyId)
+            .then((res) => recoveryData = res.recoveryData)
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/account/reset/recoveryKeys',
+      config: {
+        auth: {
+          strategy: 'accountResetToken'
+        },
+        validate: {
+          query: {
+            keys: isA.boolean().optional()
+          },
+          payload: {
+            authPW: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            wrapKb: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            recoveryKeyId: isA.string().max(256).required(),
+          }
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Password.changeRecoveryKeys', request)
+        const authPW = request.payload.authPW
+        const wrapKb = request.payload.wrapKb
+        const recoveryKeyId = request.payload.recoveryKeyId
+        const uid = request.auth.credentials.uid
+
+        let authSalt, password, verifyHash, sessionToken, keyFetchToken, account
+
+        checkRecoveryKey()
+          .then(getAccount)
+          .then(passwordReset)
+          .then(deleteRecoveryKey)
```

Comment:
Should deleting the recovery key be done last just in case there is some unexpected problem with any of the subsequent steps? If we delete the recovery key and there is a problem in, e.g., creating the sessionToken, then the user will be locked out of their account. If the delete happens last, at least the user would be able to try again and not be locked out of their account.